### DISCUSSION TITLE
correct way to build an SPA

### DISCUSSION THREAD

**Discussion by arechnitzer** [elimintz/justpy#283] [Q&A] [answered]
Really enjoying using justpy. It is fantastic and great to see so much effort gone into the docs.

Trying to build a simple SPA as a front end to a bunch of scripts we are using and justpy is perfect.

I'm wondering how you would recommend laying out my code to achieve this. A minimal example would be 
* a "landing / dashboard" view
* an "upload a configuration file + parse it" view which also report any errors in the config file
* once the config file validates bounce back to the (updated) dashboard view

How should I bulid this in justpy? At present I have a webpage (global variable) which various functions modify.
* a make_main_page function which build the webpage and its nav-bar etc and sets up a content-area div
* various "render" functions which calls delete_components on the content-area div and then rebuild the desired components for that particular view.
* the buttons in the nav bar call those render functions.

Or am I better off with a sequence of pages and redirects?

**Comment by elimintz** [comment, marked as answer]
Thanks for using JustPy!

One of the advantages of an SPA is that only the content that a user needs is changed without needing the whole page to be loaded. The first approach will give you that benefit and may be preferable.  That is the approach I use when I just want to update some content on a page based on user input.

**Reply by arechnitzer** [reply]
Many thanks for the very quick response! 

Do you have a example I could build from?

**Reply by elimintz** [reply]
Not sure this is something to build on, but it does demonstrate the concepts:
https://github.com/elimintz/justpy/blob/master/examples/dogs.py

It is a simple Quasar based SPA that displays pics of dogs of the breed selected by the user. I can find no better use for the Internet. 😉

**Reply by arechnitzer** [reply]
well - cat photos would be better (imho).

Thanks for the pointer - I'll dig. Also looking at the discussion in #225 - feels relevant somehow.

**Reply by arechnitzer** [reply]
btw (at the risk of going off topic) - what IDE do you use to write your nice quasar html?

**Reply by elimintz** [reply]
I use PyCharm. It allows you to inject other languages into strings: https://www.jetbrains.com/help/pycharm/using-language-injections.html
